### Q1. A company conducted a survey of its employees and found that 70% of the employees use the company's health insurance plan, while 40% of the employees who use the plan are smokers. What is the probability that an employee is a smoker given that he/she uses the health insurance plan?

he probability that an employee is a smoker given that he/she uses the health insurance plan is 
0.40 or 40%.

### Q2.What is the difference between Bernoulli Naive Bayes and Multinomial Naive Bayes?

- Bernoulli Naive Bayes:

Assumes binary features: Each feature can only take on two values, typically represented by 0 and 1.

Treats features as presence/absence indicators: A value of 1 indicates the feature is present, while 0 indicates it's absent.
Ignores the frequency or count of the features.

Used when features are naturally binary or binarized (e.g., text classification by word presence/absence).

- Multinomial Naive Bayes:

Assumes discrete features: Features can take on a finite number of discrete values, but not limited to just 0 and 1.

Considers the frequency or count of each feature value.

Useful for cases where feature occurrence matters, like word counts in documents, or presence of different events.

## Q3. How does Bernoulli Naive Bayes handle missing values?

- Imputation:

You might choose to impute missing values with the most common value (mode) of the feature. If 0 is more common, you could impute missing values as 0; if 1 is more common, you could impute missing values as 1.

Another option is to use a more sophisticated imputation technique, such as k-nearest neighbors imputation, to estimate the missing values based on other similar instances.

## Q4. Can Gaussian Naive Bayes be used for multi-class classification?

Yes, Gaussian Naive Bayes can be used for multi-class classification. Gaussian Naive Bayes is an extension of the Naive Bayes algorithm that assumes that continuous features follow a Gaussian (normal) distribution within each class. While it is often used for binary classification problems, it can also be extended to handle multiple classes.

### Q5.Assignment

In [1]:
pip install ucimlrepo

Note: you may need to restart the kernel to use updated packages.


In [3]:
from ucimlrepo import fetch_ucirepo 
  
# fetch dataset 
spambase = fetch_ucirepo(id=94) 
  
# data (as pandas dataframes) 
X = spambase.data.features 
y = spambase.data.targets 
  
# metadata 
print(spambase.metadata) 


{'uci_id': 94, 'name': 'Spambase', 'repository_url': 'https://archive.ics.uci.edu/dataset/94/spambase', 'data_url': 'https://archive.ics.uci.edu/static/public/94/data.csv', 'abstract': 'Classifying Email as Spam or Non-Spam', 'area': 'Computer Science', 'tasks': ['Classification'], 'characteristics': ['Multivariate'], 'num_instances': 4601, 'num_features': 57, 'feature_types': ['Integer', 'Real'], 'demographics': [], 'target_col': ['Class'], 'index_col': None, 'has_missing_values': 'no', 'missing_values_symbol': None, 'year_of_dataset_creation': 1999, 'last_updated': 'Mon Aug 28 2023', 'dataset_doi': '10.24432/C53G6X', 'creators': ['Mark Hopkins', 'Erik Reeber', 'George Forman', 'Jaap Suermondt'], 'intro_paper': None, 'additional_info': {'summary': 'The "spam" concept is diverse: advertisements for products/web sites, make money fast schemes, chain letters, pornography...\n\nThe classification task for this dataset is to determine whether a given email is spam or not.\n\t\nOur collecti

In [4]:
# variable information 
print(spambase.variables) 

                          name     role        type demographic  \
0               word_freq_make  Feature  Continuous        None   
1            word_freq_address  Feature  Continuous        None   
2                word_freq_all  Feature  Continuous        None   
3                 word_freq_3d  Feature  Continuous        None   
4                word_freq_our  Feature  Continuous        None   
5               word_freq_over  Feature  Continuous        None   
6             word_freq_remove  Feature  Continuous        None   
7           word_freq_internet  Feature  Continuous        None   
8              word_freq_order  Feature  Continuous        None   
9               word_freq_mail  Feature  Continuous        None   
10           word_freq_receive  Feature  Continuous        None   
11              word_freq_will  Feature  Continuous        None   
12            word_freq_people  Feature  Continuous        None   
13            word_freq_report  Feature  Continuous        Non

In [7]:
X.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0             0.00            0.00  ...                   0.0         0.00   
1             0.00            0.94  ...                   0.0         0.00   
2             0.64            0.25  ...                   0.0         0.01   
3             0.31            0.63  ...                   0.0         0.00   
4             0.31            0.63  ...                   0.0         0.00   

   char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0        0.000          0.0        0.778        0.000        0.000   
1        0.132          0.0        0.372        0.180        0.048   
2        0.143          0.0        0.276        0.184        0.010   
3        0.137          0.0        0.137        0.000        0.000   
4        0.135          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  
0                       278  
1                      1028  
2                      2259  
3                       191  
4                       191  

[5 rows x 57 columns]

In [8]:
y.head()

Class
0      1
1      1
2      1
3      1
4      1

In [18]:
from sklearn.naive_bayes import BernoulliNB, MultinomialNB, GaussianNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import cross_val_score


In [12]:
bernoulli_nb = BernoulliNB()
multinomial_nb = MultinomialNB()
gaussian_nb = GaussianNB()


In [15]:
# Define 10-fold cross-validation

cv = StratifiedKFold(n_splits=10, shuffle=True, random_state=42)


In [16]:
# Evaluate classifiers
def evaluate_classifier(classifier, X, y):
    accuracy = cross_val_score(classifier, X, y, cv=cv, scoring='accuracy')
    precision = cross_val_score(classifier, X, y, cv=cv, scoring='precision_macro')
    recall = cross_val_score(classifier, X, y, cv=cv, scoring='recall_macro')
    f1 = cross_val_score(classifier, X, y, cv=cv, scoring='f1_macro')
    return accuracy, precision, recall, f1


In [19]:
# Evaluate Bernoulli Naive Bayes
accuracy_b, precision_b, recall_b, f1_b = evaluate_classifier(bernoulli_nb, X, y)

# Evaluate Multinomial Naive Bayes
accuracy_m, precision_m, recall_m, f1_m = evaluate_classifier(multinomial_nb, X, y)

# Evaluate Gaussian Naive Bayes
accuracy_g, precision_g, recall_g, f1_g = evaluate_classifier(gaussian_nb, X, y)


/opt/conda/lib/python3.10/site-packages/sklearn/utils/validation.py:1141: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/opt/conda/lib/python3.10/site-packages/sklearn/utils/validation.py:1141: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/opt/conda/lib/python3.10/site-packages/sklearn/utils/validation.py:1141: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/opt/conda/lib/python3.10/site-packages/sklearn/utils/validation.py:1141: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for exam

In [21]:
# Print results
print("Bernoulli Naive Bayes:")
print(f"Accuracy: {accuracy_b.mean():.4f}")
print(f"Precision: {precision_b.mean():.4f}")
print(f"Recall: {recall_b.mean():.4f}")
print(f"F1 Score: {f1_b.mean():.4f}")
print("\n")

print("Multinomial Naive Bayes:")
print(f"Accuracy: {accuracy_m.mean():.4f}")
print(f"Precision: {precision_m.mean():.4f}")
print(f"Recall: {recall_m.mean():.4f}")
print(f"F1 Score: {f1_m.mean():.4f}")
print("\n")

print("Gaussian Naive Bayes:")
print(f"Accuracy: {accuracy_g.mean():.4f}")
print(f"Precision: {precision_g.mean():.4f}")
print(f"Recall: {recall_g.mean():.4f}")
print(f"F1 Score: {f1_g.mean():.4f}")


Bernoulli Naive Bayes:
Accuracy: 0.8857
Precision: 0.8858
Recall: 0.8735
F1 Score: 0.8785


Multinomial Naive Bayes:
Accuracy: 0.7903
Precision: 0.7813
Recall: 0.7782
F1 Score: 0.7794


Gaussian Naive Bayes:
Accuracy: 0.8203
Precision: 0.8313
Recall: 0.8443
F1 Score: 0.8195


- Discussion:

Performance Metrics:

Bernoulli Naive Bayes achieved the highest accuracy (88.57%), precision (88.58%), and F1 score (87.85%) among the three variants.

Multinomial Naive Bayes has slightly lower performance across all metrics compared to Bernoulli Naive Bayes.

Gaussian Naive Bayes performed well but didn't outperform Bernoulli Naive Bayes.

Observations:

The higher performance of Bernoulli Naive Bayes might be attributed to the nature of the dataset. If the features are more binary in nature, Bernoulli Naive Bayes, which assumes binary features, may be a better fit.

Multinomial Naive Bayes has lower performance, possibly because it's more suited for datasets with discrete features representing counts or frequencies.

Gaussian Naive Bayes assumes features follow a Gaussian distribution. Despite performing reasonably well, it might not be the best choice if the features significantly deviate from a normal distribution.

Limitations of Naive Bayes:

Assumption of Feature Independence: Naive Bayes assumes that features are conditionally independent given the class. In real-world data, this assumption may not always hold, which can affect the model's performance.

Sensitivity to Feature Distributions: Each variant of Naive Bayes makes specific assumptions about the distribution of features. If these assumptions are violated, the model's accuracy may be impacted.

Difficulty Handling Missing Values: Naive Bayes doesn't handle missing values well, and appropriate preprocessing or imputation is required.

- Conclusion:

In this specific case, Bernoulli Naive Bayes performed the best among the three variants. However, the choice of the best variant depends on the characteristics of the dataset. It's crucial to consider the nature of features and their distribution when selecting a Naive Bayes variant.

Suggestions for Future Work:

Feature Engineering: Explore different ways to represent and engineer features to potentially improve the performance of all models.

Hyperparameter Tuning: Experiment with hyperparameter tuning for each Naive Bayes variant to find the optimal settings.

Ensemble Methods: Consider ensemble methods like Random Forest or Gradient Boosting to see if they can provide better predictive performance.

Evaluate Other Models: Compare Naive Bayes models with other classification algorithms like Support Vector Machines, Decision Trees, or Neural Networks to identify the most suitable model for the dataset.

Handling Imbalanced Data: If the dataset is imbalanced, implement techniques like oversampling or undersampling to address class imbalances and evaluate their impact on model performance.